docs/testing.md

```diff
@@ -0,0 +1,454 @@
+
+async def test_idempotent_scheduling(
+    test_docket: Docket, test_worker: Worker, mock_task: AsyncMock
+):
+    """Test that tasks with same key don't duplicate."""
+    key = "unique-task-key"
+
+    # Schedule same task multiple times with same key
+    await test_docket.add(mock_task, key=key)("data1")
+    await test_docket.add(mock_task, key=key)("data2")  # Should replace
+    await test_docket.add(mock_task, key=key)("data3")  # Should replace
+
+    await test_worker.run_until_finished()
+
+    # Should only execute once with the last data
+    mock_task.assert_awaited_once_with("data3")
+```
+
+### Time-based Testing Fixtures
+
+For testing scheduled tasks, add a time fixture:
+
+```python
+from datetime import datetime, timezone
+from functools import partial
+
+@pytest.fixture
+def now():
+    """Consistent time source for tests."""
+    return partial(datetime.now, timezone.utc)
+
+async def test_scheduled_execution(test_docket, test_worker, mock_task, now):
+    """Test that tasks execute at the right time."""
+    # Schedule for 100ms in the future
+    when = now() + timedelta(milliseconds=100)
+    await test_docket.add(mock_task, when=when)("scheduled data")
+
+    await test_worker.run_until_finished()
+
+    # Verify it ran and that enough time passed
+    mock_task.assert_awaited_once_with("scheduled data")
+    assert now() >= when
+```
+
+## Running Until Finished
+
+For tests and batch processing, use `run_until_finished()` to process all pending tasks then stop:
+
+```python
+async def test_order_processing():
+    async with Docket() as docket:
+        docket.register(process_order)
+        docket.register(send_confirmation)
+        docket.register(update_inventory)
+
+        # Schedule some work
+        await docket.add(process_order)(order_id=123)
+        await docket.add(send_confirmation)(order_id=123)
+        await docket.add(update_inventory)(product_id=456)
+
+        # Process all pending tasks
+        async with Worker(docket) as worker:
+            await worker.run_until_finished()
+
+        # Now verify results
+        assert order_is_processed(123)
+        assert confirmation_was_sent(123)
+        assert inventory_was_updated(456)
+```
+
+This works well for testing workflows where you need to ensure all tasks complete before making assertions.
+
+### Testing Task Registration
+
+Test that tasks are properly registered and can be called by name:
+
+```python
+async def test_task_registration_by_name(test_docket, test_worker, mock_task):
+    """Test executing tasks by string name."""
+    test_docket.register(mock_task)
+
+    # Execute by name instead of function reference
+    await test_docket.add("mock_task")("test data")
+
+    await test_worker.run_until_finished()
+
+    mock_task.assert_awaited_once_with("test data")
+```
+
+## Controlling Perpetual Tasks
+
+Use `run_at_most()` to limit how many times specific tasks run, which is essential for testing perpetual tasks:
```

Comment:
Can we make all function references like this links to the appropriate API reference?